microcosm_elasticsearch/models.py

```diff
@@ -3,17 +3,56 @@
 
 """
 from elasticsearch_dsl import Date, DocType, Keyword
+from elasticsearch_dsl.document import DocTypeOptions, DocTypeMeta
 
 
-class Model(DocType):
+class ModelMeta(DocTypeMeta):
+    def __new__(cls, name, bases, attrs):
+        attrs["_doc_type"] = ModelOptions(name, bases, attrs)
+        return super(DocTypeMeta, cls).__new__(cls, name, bases, attrs)
+
+
+class ModelOptions(DocTypeOptions):
+    def __init__(self, name, bases, attrs):
+        if "__doctype_field__" in attrs:
+            # Dynamically set the field for polymorphic doctype
+            doctype_field_name = attrs["__doctype_field__"]
+            attrs[doctype_field_name] = Keyword(required=True)
+        super().__init__(name, bases, attrs)
+
+
+class ModelBase(DocType):
     """
     Every persistent entity should have a primary key id and created/updated timestamps.
 
+    `ModelBase` has all the behavior we want from the `Model` base class; but since it defines
+    a __doctype_field__ we want it to avoid the special behavior defined in `ModelMeta`. This is why
+    `Model` inherits from this class and has the `ModelMeta` metaclass.
+
     """
+    # By default when creating a document from a `Model` subclass instance
+    # Its doctype will be set to the lowercased subclass name
+    # To override this declare:
+    # __doctype_name__ = "<other name>"
+
+    # The following will set up a field to hold the document's polymorphic doctype
+    # By default the field name is "doctype", override in child classes to change this
+    # Note that if you change it here you should also override the SearchIndex's `mapping_type_name` property
+    __doctype_field__ = "doctype"
```

Comment:
Concretely speaking, if I define
```class SomeModel(Model):
    __doctype_name__ = "the_some_model"
    __doctype_field__ = "the_doctype:
```

Then an ES document created from a `SomeModel` instance would bear the field/value pair:
```
{
    "the_doctype": "the_some_model",
}
```
And the value of `the_doctype` will be used to instantiate the right class from that ES result.

Reply:
This would be good to have in the doc string...

Reply:
I added an explanation in the `README` with a reference to that in the docstring.